File: sdks/python/linktor/types/vre.py
"""
VRE (Visual Response Engine) types
"""

from enum import Enum
from typing import Any, Optional
from pydantic import BaseModel, Field


class VREOutputFormat(str, Enum):
    """Output format for rendered images"""

    PNG = "png"
    WEBP = "webp"
    JPEG = "jpeg"


class VREChannelType(str, Enum):
    """Channel type for VRE rendering"""

    WHATSAPP = "whatsapp"
    TELEGRAM = "telegram"
    WEB = "web"
    EMAIL = "email"


class VRETemplateType(str, Enum):
    """Available template types"""

    MENU_OPCOES = "menu_opcoes"
    CARD_PRODUTO = "card_produto"
    STATUS_PEDIDO = "status_pedido"
    LISTA_PRODUTOS = "lista_produtos"
    CONFIRMACAO = "confirmacao"
    COBRANCA_PIX = "cobranca_pix"


class OrderStatus(str, Enum):
    """Order status for status_pedido template"""

    RECEBIDO = "recebido"
    SEPARACAO = "separacao"
    FATURADO = "faturado"
    TRANSPORTE = "transporte"
    ENTREGUE = "entregue"


class StockStatus(str, Enum):
    """Stock status for products"""

    DISPONIVEL = "disponivel"
    BAIXO = "baixo"
    INDISPONIVEL = "indisponivel"


# ============================================
# Render Request/Response
# ============================================


class VRERenderRequest(BaseModel):
    """Request to render a VRE template"""

    tenant_id: str = Field(alias="tenantId")
    template_id: VRETemplateType = Field(alias="templateId")
    data: dict[str, Any]
    channel: Optional[VREChannelType] = None
    format: Optional[VREOutputFormat] = None
    width: Optional[int] = None
    quality: Optional[int] = None
    scale: Optional[float] = None

    class Config:
        populate_by_name = True


class VRERenderResponse(BaseModel):
    """Response from rendering a template"""

    image_base64: str = Field(alias="imageBase64")
    caption: str
    width: int
    height: int
    format: VREOutputFormat
    render_time_ms: int = Field(alias="renderTimeMs")
    size_bytes: Optional[int] = Field(None, alias="sizeBytes")
    cache_hit: Optional[bool] = Field(None, alias="cacheHit")

    class Config:
        populate_by_name = True


# ============================================
# Render and Send
# ============================================


class VRERenderAndSendRequest(BaseModel):
    """Request to render a template and send it directly to a conversation"""

    conversation_id: str = Field(alias="conversationId")
    template_id: VRETemplateType = Field(alias="templateId")
    data: dict[str, Any]
    caption: Optional[str] = None
    follow_up_text: Optional[str] = Field(None, alias="followUpText")

    class Config:
        populate_by_name = True


class VRERenderAndSendResponse(BaseModel):
    """Response from render and send"""

    message_id: str = Field(alias="messageId")
    image_url: str = Field(alias="imageUrl")
    caption: str

    class Config:
        populate_by_name = True


# ============================================
# Templates
# ============================================


class VRETemplate(BaseModel):
    """VRE template definition"""

    id: VRETemplateType
    name: str
    description: str
    schema: dict[str, Any]


class VREListTemplatesResponse(BaseModel):
    """Response from listing templates"""

    templates: list[VRETemplate]


class VREPreviewRequest(BaseModel):
    """Request to preview a template"""

    template_id: VRETemplateType = Field(alias="templateId")
    data: Optional[dict[str, Any]] = None

    class Config:
        populate_by_name = True


class VREPreviewResponse(BaseModel):
    """Response from template preview"""

    image_base64: str = Field(alias="imageBase64")
    width: int
    height: int

    class Config:
        populate_by_name = True


# ============================================
# Template Data Types
# ============================================


class MenuOpcaoData(BaseModel):
    """Menu option for menu_opcoes template"""

    label: str
    descricao: Optional[str] = None
    icone: Optional[str] = None


class MenuOpcoesData(BaseModel):
    """Data for menu_opcoes template"""

    titulo: str
    subtitulo: Optional[str] = None
    opcoes: list[MenuOpcaoData]
    mensagem_antes: Optional[str] = Field(None, alias="mensagemAntes")

    class Config:
        populate_by_name = True


class CardProdutoData(BaseModel):
    """Data for card_produto template"""

    nome: str
    sku: Optional[str] = None
    preco: float
    unidade: str
    estoque: Optional[int] = None
    imagem_url: Optional[str] = Field(None, alias="imagemUrl")
    destaque: Optional[str] = None
    mensagem: Optional[str] = None

    class Config:
        populate_by_name = True


class StatusPedidoStep(BaseModel):
    """Timeline step for status_pedido template"""

    status: str
    icon: str
    label: str


class StatusPedidoData(BaseModel):
    """Data for status_pedido template"""

    numero_pedido: str = Field(alias="numeroPedido")
    status_atual: OrderStatus = Field(alias="statusAtual")
    itens_resumo: Optional[str] = Field(None, alias="itensResumo")
    valor_total: Optional[float] = Field(None, alias="valorTotal")
    previsao_entrega: Optional[str] = Field(None, alias="previsaoEntrega")
    motorista: Optional[str] = None
    mensagem: Optional[str] = None

    class Config:
        populate_by_name = True


class ListaProdutoItem(BaseModel):
    """Product item for lista_produtos template"""

    nome: str
    preco: float
    unidade: Optional[str] = None
    estoque_status: Optional[StockStatus] = Field(None, alias="estoqueStatus")
    sku: Optional[str] = None
    emoji: Optional[str] = None

    class Config:
        populate_by_name = True


class ListaProdutosData(BaseModel):
    """Data for lista_produtos template"""

    titulo: str
    produtos: list[ListaProdutoItem]
    mensagem: Optional[str] = None


class ConfirmacaoItem(BaseModel):
    """Confirmation item for confirmacao template"""

    nome: str
    quantidade: Optional[str] = None
    preco: float
    emoji: Optional[str] = None


class ConfirmacaoData(BaseModel):
    """Data for confirmacao template"""

    titulo: Optional[str] = None
    subtitulo: Optional[str] = None
    itens: list[ConfirmacaoItem]
    valor_total: float = Field(alias="valorTotal")
    previsao_entrega: Optional[str] = Field(None, alias="previsaoEntrega")
    mensagem: Optional[str] = None

    class Config:
        populate_by_name = True


class CobrancaPixData(BaseModel):
    """Data for cobranca_pix template"""

    valor: float
    numero_pedido: Optional[str] = Field(None, alias="numeroPedido")
    pix_payload: str = Field(alias="pixPayload")
    expiracao: Optional[str] = None
    mensagem: Optional[str] = None

    class Config:
        populate_by_name = True


__all__ = [
    "VREOutputFormat",
    "VREChannelType",
    "VRETemplateType",
    "OrderStatus",
    "StockStatus",
    "VRERenderRequest",
    "VRERenderResponse",
    "VRERenderAndSendRequest",
    "VRERenderAndSendResponse",
    "VRETemplate",
    "VREListTemplatesResponse",
    "VREPreviewRequest",
    "VREPreviewResponse",
    "MenuOpcaoData",
    "MenuOpcoesData",
    "CardProdutoData",
    "StatusPedidoStep",
    "StatusPedidoData",
    "ListaProdutoItem",
    "ListaProdutosData",
    "ConfirmacaoItem",
    "ConfirmacaoData",
    "CobrancaPixData",
]

File: sdks/rust/src/types/vre.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Output format for rendered images
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum VREOutputFormat {
    Png,
    Webp,
    Jpeg,
}

/// Channel type for VRE rendering
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum VREChannelType {
    Whatsapp,
    Telegram,
    Web,
    Email,
}

/// Available template types
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum VRETemplateType {
    MenuOpcoes,
    CardProduto,
    StatusPedido,
    ListaProdutos,
    Confirmacao,
    CobrancaPix,
}

/// Order status for status_pedido template
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum OrderStatus {
    Recebido,
    Separacao,
    Faturado,
    Transporte,
    Entregue,
}

/// Stock status for products
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum StockStatus {
    Disponivel,
    Baixo,
    Indisponivel,
}

/// Render request
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct VRERenderRequest {
    pub tenant_id: String,
    pub template_id: String,
    pub data: HashMap<String, serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub channel: Option<VREChannelType>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub format: Option<VREOutputFormat>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub width: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub quality: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub scale: Option<f64>,
}

impl VRERenderRequest {
    pub fn new(tenant_id: impl Into<String>, template_id: impl Into<String>, data: HashMap<String, serde_json::Value>) -> Self {
        Self {
            tenant_id: tenant_id.into(),
            template_id: template_id.into(),
            data,
            channel: None,
            format: None,
            width: None,
            quality: None,
            scale: None,
        }
    }

    pub fn channel(mut self, channel: VREChannelType) -> Self {
        self.channel = Some(channel);
        self
    }

    pub fn format(mut self, format: VREOutputFormat) -> Self {
        self.format = Some(format);
        self
    }

    pub fn width(mut self, width: i32) -> Self {
        self.width = Some(width);
        self
    }

    pub fn quality(mut self, quality: i32) -> Self {
        self.quality = Some(quality);
        self
    }

    pub fn scale(mut self, scale: f64) -> Self {
        self.scale = Some(scale);
        self
    }
}

/// Render response
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct VRERenderResponse {
    pub image_base64: String,
    pub caption: String,
    pub width: i32,
    pub height: i32,
    pub format: VREOutputFormat,
    pub render_time_ms: i32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub size_bytes: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cache_hit: Option<bool>,
}

/// Render and send request
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct VRERenderAndSendRequest {
    pub conversation_id: String,
    pub template_id: String,
    pub data: HashMap<String, serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub caption: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub follow_up_text: Option<String>,
}

impl VRERenderAndSendRequest {
    pub fn new(conversation_id: impl Into<String>, template_id: impl Into<String>, data: HashMap<String, serde_json::Value>) -> Self {
        Self {
            conversation_id: conversation_id.into(),
            template_id: template_id.into(),
            data,
            caption: None,
            follow_up_text: None,
        }
    }

    pub fn caption(mut self, caption: impl Into<String>) -> Self {
        self.caption = Some(caption.into());
        self
    }

    pub fn follow_up_text(mut self, follow_up_text: impl Into<String>) -> Self {
        self.follow_up_text = Some(follow_up_text.into());
        self
    }
}

/// Render and send response
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct VRERenderAndSendResponse {
    pub message_id: String,
    pub image_url: String,
    pub caption: String,
}

/// Template definition
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct VRETemplate {
    pub id: String,
    pub name: String,
    pub description: String,
    pub schema: HashMap<String, serde_json::Value>,
}

/// List templates response
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct VREListTemplatesResponse {
    pub templates: Vec<VRETemplate>,
}

/// Preview request
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct VREPreviewRequest {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data: Option<HashMap<String, serde_json::Value>>,
}

impl VREPreviewRequest {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn data(mut self, data: HashMap<String, serde_json::Value>) -> Self {
        self.data = Some(data);
        self
    }
}

/// Preview response
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct VREPreviewResponse {
    pub image_base64: String,
    pub width: i32,
    pub height: i32,
}

/// Menu option for menu_opcoes template
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MenuOpcaoData {
    pub label: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub descricao: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub icone: Option<String>,
}

impl MenuOpcaoData {
    pub fn new(label: impl Into<String>) -> Self {
        Self {
            label: label.into(),
            descricao: None,
            icone: None,
        }
    }

    pub fn descricao(mut self, descricao: impl Into<String>) -> Self {
        self.descricao = Some(descricao.into());
        self
    }

    pub fn icone(mut self, icone: impl Into<String>) -> Self {
        self.icone = Some(icone.into());
        self
    }
}

/// Product card data
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct CardProdutoData {
    pub nome: String,
    pub preco: f64,
    pub unidade: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sku: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub estoque: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub imagem_url: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub destaque: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub mensagem: Option<String>,
}

impl CardProdutoData {
    pub fn new(nome: impl Into<String>, preco: f64, unidade: impl Into<String>) -> Self {
        Self {
            nome: nome.into(),
            preco,
            unidade: unidade.into(),
            sku: None,
            estoque: None,
            imagem_url: None,
            destaque: None,
            mensagem: None,
        }
    }
}

/// Order status data
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct StatusPedidoData {
    pub numero_pedido: String,
    pub status_atual: OrderStatus,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub itens_resumo: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub valor_total: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub previsao_entrega: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub motorista: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub mensagem: Option<String>,
}

impl StatusPedidoData {
    pub fn new(numero_pedido: impl Into<String>, status_atual: OrderStatus) -> Self {
        Self {
            numero_pedido: numero_pedido.into(),
            status_atual,
            itens_resumo: None,
            valor_total: None,
            previsao_entrega: None,
            motorista: None,
            mensagem: None,
        }
    }
}

/// Product list item
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct ListaProdutoItem {
    pub nome: String,
    pub preco: f64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub unidade: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub estoque_status: Option<StockStatus>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sku: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub emoji: Option<String>,
}

impl ListaProdutoItem {
    pub fn new(nome: impl Into<String>, preco: f64) -> Self {
        Self {
            nome: nome.into(),
            preco,
            unidade: None,
            estoque_status: None,
            sku: None,
            emoji: None,
        }
    }
}

/// Confirmation item
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ConfirmacaoItem {
    pub nome: String,
    pub preco: f64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub quantidade: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub emoji: Option<String>,
}

impl ConfirmacaoItem {
    pub fn new(nome: impl Into<String>, preco: f64) -> Self {
        Self {
            nome: nome.into(),
            preco,
            quantidade: None,
            emoji: None,
        }
    }
}

/// PIX payment data
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub struct CobrancaPixData {
    pub valor: f64,
    pub pix_payload: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub numero_pedido: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub expiracao: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub mensagem: Option<String>,
}

impl CobrancaPixData {
    pub fn new(valor: f64, pix_payload: impl Into<String>) -> Self {
        Self {
            valor,
            pix_payload: pix_payload.into(),
            numero_pedido: None,
            expiracao: None,
            mensagem: None,
        }
    }
}
